CUDACachingAllocator: Keep one event queue per stream

nelhage opened the pull request:
Fixes #71616

This fixes the leaks in my test case. I have not tested it on our big models yet, but will report back if we can.

This potentially impacts allocator performance in that it slightly increases the amount of CPU memory we allocate for data structures, and it means that `process_events` may look at a larger number of events in the case where there are multiple streams with long-running ops on them.

However, I suspect that in general, either:
- An application isn't using very many streams or very many long-running ops, in which case the performance is essentially the same
- Or, they are, which is precisely the case where #71616 bites you, and so freeing memory faster is probably more valuable than the slight CPU overhead here.

I'm not attached to this approach or any of its details, but figured it was worth throwing up for discussion.

nelhage committed:
CUDACachingAllocator: Keep one event queue per stream

nelhage committed:
Comment update

colesbury reviewed (commented):
This looks good to me. I am convinced by the argument in the PR that the change is worth the (small) CPU cost.

ngimel commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -404,7 +404,7 @@ class DeviceCachingAllocator {
   std::vector<Block*> needs_events_deferred_until_no_capture;
   // outstanding cuda events
-  std::deque<std::pair<cudaEvent_t, Block*>> cuda_events;
+  std::unordered_map<cuda::CUDAStream, std::deque<std::pair<cudaEvent_t, Block*>>> cuda_events;
```
use `ska::flat_hash_map` here, it's faster?

nelhage replied:
Will change! I deferred to the fact that this code was already using STL maps elsewhere, but I'm always happy to use a properly-optimized hash map instead of the STL.

nelhage replied:
Should I change the other maps in this file for consistency while I'm here, or leave that for a followup PR?

ngimel reviewed (commented):
This approach looks good to me too, I have one small comment.

nelhage commented:
> This fixes the leaks in my test case. I have not tested it on our big models yet, but will report back if we can.
We've since run some tests on our training runs, and this gets back a bit more than half of the [active - allocated] gap, at no measurable performance regression (per our metrics). So this is definitely a substantial win for us.

I'm pursuing some other theories for the rest of the gap, including that some of it is the system working-as-intended because we're freeing tensors that are still in-flight; I'll update the other issue or open new issues if I find other things that I think want to be fixed on the PyTorch side.

nelhage committed:
CUDACachingAllocator: Switch to skarupke maps everywhere.

nelhage commented:
I chose to update all the maps and sets to `ska::` for consistency. Happy to back that out if you'd prefer a second PR.

ngimel commented:
That's fine, there's #71667 to do that, but we can do it in this PR also.

colesbury commented:
Does `ska::flat_hash_map` have the same guarantees about iterator validity? In particular does `erase()` only invalidate the iterator to the delete entry (and not other iterators)?

nelhage commented:
> Does ska::flat_hash_map have the same guarantees about iterator validity? In particular does erase() only invalidate the iterator to the delete entry (and not other iterators)?

I had this question. Frustratingly `ska::flat_hash_map` does not seem to explicitly document the invalidation policy.

(`abseil`'s flat_hash_map does document this detail and `erase` promises not to rehash, but that's not the one we're using).

Looking at [the source code](https://github.com/pytorch/pytorch/blob/e755a4f12456f8743a4b0bb786254dd37102a8ef/c10/util/flat_hash_map.h#L654-L669), I'm not familiar enough with `ska::flat_hash_map` offhand to say for sure, but it does worry me that it moves elements around. The safest course of action is probably to rely on the return value from `erase` instead. so I'll put that change up.

nelhage committed:
Use the return value from ska::flat_hash_map::erase(iterator).

Instead of relying on erase() having the same invalidation guarantees
as std::unordered_map.

nelhage commented:
Seems like CI completely lost the plot on that push I don't understand why. Can someone kick it (if that's the right course of action)?

ngimel commented:
github.com is having a bad day, so lots of jobs are failing

ngimel commented:
Sorry about that @nelhage, but your rebase fell on a time when master was broken from an unrelated PR (hence test failures on yours), can you please rebase again? Really sorry about it, we try to keep CI in good state, but it doesn't always work.

nelhage commented:
Done! I'm in awe of how much CI y'all have, and I know the struggles of keeping it green even for much simpler builds :)

mcarilli commented:
I think it's still possible to hit head of line blocking behavior with this PR's diffs (ie, even within one stream):
```python
a = torch.tensor(...)
b = torch.tensor(...)
with torch.cuda.stream(s):
    a.record_stream(s)
    b.record_stream(s)
    quick_usage(a)
    long_usage(b)
    del b
    del a
```
b's end of life events enter the per-stream deques before a's, but take a long time to resolve, delaying the reusability of a.

In any case this PR is better than the status quo.

A possible wild idea I had was to avoid end-of-life event deques entirely. Instead, blocks with recorded streams and associated end-of-life events at "free()" time go into a dedicated pool, ie "small/largeBlocksMayStillHaveEventsPending". This pool would be used as an intermediate fallback by malloc, ie, malloc would service allocation requests by first look for a size-appropriate block in the ordinary pools, then if that fails look for a size-appropriate block in small/largeBlocksMayStillHaveEventsPending. If suitable block found, check its events. If events resolved, use the block and delete it from its Pending pool. If not, move to the next biggest block and check its events if desired, or bail out to the usual final fallbacks (try cudaMalloc, then error). A block returned from the Pending pool to service an allocation would effectively be reset (losing its Pending-specific identity) (which is fine because it'll only be returned if its events have resolved).

One downside of this approach is the intermediate-pool fallback would be another std::set lookup.
